Mathlib/CategoryTheory/Limits/Presentation.lean

```diff
@@ -0,0 +1,204 @@
+
+## TODOs:
+
+- Dualise to obtain `LimitPresentation`.
+-/
+
+universe s t w v u
+
+namespace CategoryTheory.Limits
+
+variable {C : Type u} [Category.{v} C]
+
+/-- A colimit presentation of `X` over `J` is a diagram `{Dᵢ}` in `C` and natural maps
+`sᵢ : Dᵢ ⟶ X` making `X` into the colimit of the `Dᵢ`. -/
+structure ColimitPresentation (J : Type w) [Category.{t} J] (X : C) where
+  /-- The diagram `{Dᵢ}`. -/
+  diag : J ⥤ C
+  /-- The natural maps `sᵢ : Dᵢ ⟶ X`. -/
+  natTrans : diag ⟶ (Functor.const J).obj X
+  /-- `X` is the colimit of the `Dᵢ` via `sᵢ`. -/
+  isColimit : IsColimit (Cocone.mk _ natTrans)
+
+variable {J : Type w} [Category.{t} J] {X : C}
+
+namespace ColimitPresentation
+
+initialize_simps_projections ColimitPresentation (-isColimit)
+
+/-- The cocone associated to a colimit presentation. -/
+abbrev cocone (pres : ColimitPresentation J X) : Cocone pres.diag :=
+  Cocone.mk _ pres.natTrans
+
+/-- The canonical colimit presentation of any object over a point. -/
+@[simps]
+noncomputable
+def self (X : C) : ColimitPresentation PUnit.{s + 1} X where
+  diag := (Functor.const _).obj X
+  natTrans := 𝟙 _
+  isColimit := isColimitConstCocone _ _
+
+/-- If `F` preserves colimits of shape `J`, it maps colimit presentations of `X` to
+colimit presentations of `F(X)`. -/
+@[simps]
+noncomputable
+def map (P : ColimitPresentation J X) {D : Type*} [Category D] (F : C ⥤ D)
+    [PreservesColimitsOfShape J F] : ColimitPresentation J (F.obj X) where
+  diag := P.diag ⋙ F
+  natTrans := Functor.whiskerRight P.natTrans F ≫ (F.constComp _ _).hom
+  isColimit := by
+    convert isColimitOfPreserves F P.isColimit
+    ext j
+    simp
+
+/-- Map a colimit presentation under an isomorphism. -/
+@[simps]
+def ofIso (P : ColimitPresentation J X) {Y : C} (e : X ≅ Y) : ColimitPresentation J Y where
+  diag := P.diag
+  natTrans := P.natTrans ≫ (Functor.const J).map e.hom
+  isColimit := P.isColimit.ofIsoColimit (Cocones.ext e fun _ ↦ rfl)
+
+section
+
+variable {J : Type*} {I : J → Type*} [Category J] [∀ j, Category (I j)]
+  {D : J ⥤ C} {P : ∀ j, ColimitPresentation (I j) (D.obj j)}
+
+set_option linter.unusedVariables false in
+/-- The type underlying the category used in the construction of the composition
+of colimit presentations. This is simply `Σ j, I j` but with a different category structure. -/
+@[nolint unusedArguments]
+def Total (P : ∀ j, ColimitPresentation (I j) (D.obj j)) : Type _ :=
+  Σ j, I j
+
+variable (P) in
+/-- Constructor for `Total` to guide type checking. -/
+abbrev Total.mk (i : J) (k : I i) : Total P := ⟨i, k⟩
+
+/-- Morphisms in the `Total` category. -/
+@[ext]
+structure Total.Hom (k l : Total P) where
+  /-- The underlying morphism in the first component. -/
+  base : k.1 ⟶ l.1
+  /-- A morphism in `C`. -/
+  hom : (P k.1).diag.obj k.2 ⟶ (P l.1).diag.obj l.2
+  w : (P k.1).natTrans.app k.2 ≫ D.map base = hom ≫ (P l.1).natTrans.app l.2 := by aesop_cat
```

Comment:
My understanding is that we are supposed to replace `aesop_cat` by `cat_disch` so as to test whether `grind` may replace `aesop_cat` eventually.
```suggestion
  w : (P k.1).natTrans.app k.2 ≫ D.map base = hom ≫ (P l.1).natTrans.app l.2 := by cat_disch
```